src/kademlia/dht_tracker.cpp

```diff
@@ -536,19 +536,39 @@ namespace libtorrent { namespace dht
 		return true;
 	}
 
-	namespace {
-
-	void add_node_fun(void* userdata, node_entry const& e)
+	std::vector<std::pair<node_id, udp::endpoint>> dht_tracker::live_nodes(node_id const& nid
+		, std::function<bool(sha1_hash const&, udp::endpoint const&)> cb)
 	{
-		auto v = static_cast<std::vector<udp::endpoint>*>(userdata);
-		v->push_back(e.ep());
+		std::vector<std::pair<node_id, udp::endpoint>> ret;
+
+		// TODO: figure out a better solution when multi-home is implemented
+		node const* dht = m_dht.nid() == nid ? &m_dht
+#if TORRENT_USE_IPV6
+			: m_dht6.nid() == nid ? &m_dht6 : nullptr;
+#else
+			: nullptr
+#endif
+
+		if (dht == nullptr) return ret;
+
+		dht->m_table.for_each_node([&cb, &ret](node_entry const& e)
+		{
+			udp::endpoint const endp = e.endpoint;
+			if (cb(e.id, endp)) ret.emplace_back(e.id, endp);
```

Comment:
Is this filter function important? I'm a bit hesitant, because it makes it a lot easier to screw things up and make it look like a bug in libtorrent